core/src/main/java/com/oath/oak/EntrySet.java

```diff
@@ -0,0 +1,951 @@
+    static class OpData {
+        final int entryIndex;
+        final long newValueReference;
+        long oldValueReference;
+        final int oldVersion;
+        final int newVersion;
+        final Slice slice;
+
+        OpData(int entryIndex, long oldValueReference, long newValueReference,
+            int oldVersion, int newVersion, Slice slice) {
+            this.entryIndex = entryIndex;
+            this.newValueReference = newValueReference;
+            this.oldValueReference = oldValueReference;
+            this.oldVersion = oldVersion;
+            this.newVersion = newVersion;
+            this.slice = slice;
+        }
+    }
+
+    /* LookUp is a class that encapsulates the EntrySet's information, from when entry (key) was
+     * (or wasn't) found and until the operation was completed.
+     * It should not be used by other classes, just transferred.
+     * TODO: probably to merge with OpData and create some single operation context class */
+    static class LookUp {
+        /**
+         * valueSlice is used for easier access to the off-heap memory. The location pointed by it
+         * is the one referenced by valueReference.
+         * If it is {@code null} it means the value is deleted or marked as deleted but still referenced by
+         * valueReference.
+         */
+        Slice valueSlice;
+        /**
+         * Set to true when entry is found deleted, but not yet suitable to be reused.
+         * Deletion consists of 3 steps: (1) mark off-heap deleted (LP),
+         * (2) CAS value reference to invalid, (3) CAS value version to negative.
+         * If not all three steps are done entry can not be reused for new insertion.
+         */
+        boolean isFinalizeDeletionNeeded;
+        /**
+         * valueReference is composed of 3 numbers: block ID, value position and value length. All these numbers are
+         * squashed together into a long using the VALUE masks, shifts and indices.
+         * When it equals to {@code INVALID_VALUE_REFERENCE} is means that there is no value referenced from entryIndex.
+         * This field is usually used for CAS purposes since it sits in each entry.
+         */
+        final long valueReference;
+        int entryIndex;
+        /**
+         * This is the version of the value referenced by {@code valueReference}.
+         * If {@code valueReference == INVALID_VALUE_REFERENCE}, then:
+         * {@code version <= INVALID_VERSION} if the removal was completed.
+         * {@code version > INVALID_VERSION} if the removal was not completed.
+         * else
+         * {@code version <= INVALID_VERSION} if the insertion was not completed.
+         * {@code version > INVALID_VERSION} if the insertion was completed.
+         */
+        int version;
+        /* key reference, build differently than value reference
+          * TODO: eliminate parsing key and value references by any entity other than EntrySet */
```

Comment:
if this todo is completed by this PR then remove todo

Reply:
yes, it was done, removed todo